datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java

```diff
@@ -0,0 +1,47 @@
+package com.linkedin.datahub.graphql.api;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.dataloader.DataLoaderRegistry;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Primary;
+import org.springframework.stereotype.Component;
+import org.springframework.web.context.request.WebRequest;
+
+import com.linkedin.datahub.graphql.GraphQLEngine;
+import com.linkedin.datahub.graphql.QueryContext;
+import com.linkedin.datahub.graphql.context.SpringQueryContext;
+
+import graphql.ExecutionInput;
+import graphql.ExecutionResult;
+import graphql.GraphQL;
+import graphql.spring.web.servlet.GraphQLInvocation;
+import graphql.spring.web.servlet.GraphQLInvocationData;
+
+@Component
+@Primary
+public class QueryGraphQLInvocation implements GraphQLInvocation {
+
+    @Autowired
+    GraphQL graphQL;
+
+    @Autowired
+    GraphQLEngine graphQLEngine;
+
+    public static final String APPNAME = "DatahubGmsGraphQLApp";
+
+    @Override
+    public CompletableFuture<ExecutionResult> invoke(GraphQLInvocationData invocationData, WebRequest webRequest) {
+        QueryContext queryContext = new SpringQueryContext(true, APPNAME);
+        DataLoaderRegistry register = graphQLEngine.createDataLoaderRegistry(graphQLEngine.get_dataLoaderSuppliers());
+
+        ExecutionInput executionInput = ExecutionInput.newExecutionInput()
+            .query(invocationData.getQuery())
+            .operationName(invocationData.getOperationName())
+            .variables(invocationData.getVariables())
+            .dataLoaderRegistry(register)
+            .context(queryContext)
+            .build();
+        return graphQL.executeAsync(executionInput);
```

Comment:
Are we unable to leverage the `execute` API that exists within `GraphQLEngine`? 

It seems that `GraphQLEngine` is only being used here to help initialize the DataLoaders, is that the correct interpretation?

Reply:
As shown in this example https://www.graphql-java.com/tutorials/getting-started-with-spring-boot/#schema `graphql-java-spring-boot-starter-webmvc` dependency works with the `GraphQL` bean.

Main reason to use the `GraphQLEngine` was to override the default `GraphQLContext` with `QueryContext` thereby supplying the dataProviders.

If we have to use the `GraphQLEngine` execute In my opinion i have to implement the low level details which is taken care by `graphql-java-spring-boot-starter-webmvc` (https://github.com/graphql-java/graphql-java-spring/blob/master/graphql-java-spring-webmvc/src/main/java/graphql/spring/web/servlet/components/GraphQLController.java)

Reply:
What I meant to suggest is... Why not do something like the following instead:

```
return graphQLEngine.execute(invocationData.getQuery(), invocationData.getVariables(), queryContext); 
```

If you did this, you likely wouldn't need _both_ `GraphQL` and `GraphQLEngine` in this class, just the latter.

Reply:
Got it!